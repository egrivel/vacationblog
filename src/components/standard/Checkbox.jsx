import React from 'react';
import PropTypes from 'prop-types';

import Formrow from './Formrow.jsx';

class Checkbox extends React.Component {
  static propTypes = {
    fieldId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selected: PropTypes.bool,
    onChange: PropTypes.func.isRequired
  }

  _onChange(event) {
    this.props.onChange(event.target.checked, this.props.fieldId);
  }

  _toggleValue() {
    this.props.onChange(!this.props.selected, this.props.fieldId);
  }

  render() {
    const {fieldId, label, selected} = this.props;
    return (
      <Formrow
        key={'k-' + fieldId + selected}
      >
        <label
          className="checkbox"
          htmlFor={fieldId}
          onClick={() => this._toggleValue()}
        >
          <input
            type="checkbox"
            id={fieldId}
            value="Y"
            checked={selected}
            onChange={event => this._onChange(event)}
          />
          {label}
        </label>
      </Formrow>
    );
  }
}

export default Checkbox;
