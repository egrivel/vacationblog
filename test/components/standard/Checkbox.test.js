import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import enzyme from 'enzyme';

import Checkbox from '../../../src/components/standard/Checkbox';

/*
 * Test strategy
 * The component has one optional prop (selected) and one callback
 * function (onChange). The requirements are:
 *  - the component renders differently when selected or when not selected
 *  - the onChange is fired when clicking on either the underlying
 *    input or on the rendered label
 */
describe('components/standard/Checkbox', () => {
  const testFieldId = 'test-field-id';
  const testLabel = 'Test Label';
  let onChangeStub;
  let testEvent;

  beforeEach(() => {
    onChangeStub = sinon.stub();
    testEvent = {
      preventDefault() {},
      target: {}
    };
  });

  afterEach(() => {
    sinon.restore();
  });

  describe('#render', () => {
    test('renders unchecked', () => {
      const render = renderer.create(
        <Checkbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
        />
      );
      expect(render.toJSON()).toMatchSnapshot();
    });

    test('renders checked', () => {
      const render = renderer.create(
        <Checkbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
          selected
        />
      );
      expect(render.toJSON()).toMatchSnapshot();
    });
  });

  describe('handles onChange', () => {
    it('unselected, click input', () => {
      const component = enzyme.shallow(
        <Checkbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
        />
      );
      // when the checkbox is unchecked, the event sets it to checked
      testEvent.target.checked = true;
      component.find('input').simulate('change', testEvent);
      expect(onChangeStub.callCount).toBe(1);
      // Expect that the new selection value is true
      expect(onChangeStub.args[0][0]).toBe(true);
    });

    it('unselected, click label', () => {
      const component = enzyme.shallow(
        <Checkbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
        />
      );
      component.find('label').simulate('click', testEvent);
      expect(onChangeStub.callCount).toBe(1);
      // Expect that the new selection value is true
      expect(onChangeStub.args[0][0]).toBe(true);
    });

    it('selected, click input', () => {
      const component = enzyme.shallow(
        <Checkbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
          selected
        />
      );
      // when the checkbox is checked, the event sets it to unchecked
      testEvent.target.checked = false;
      component.find('input').simulate('change', testEvent);
      expect(onChangeStub.callCount).toBe(1);
      // Expect that the new selection value is false
      expect(onChangeStub.args[0][0]).toBe(false);
    });

    it('selected, click label', () => {
      const component = enzyme.shallow(
        <Checkbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
          selected
        />
      );
      component.find('label').simulate('click', testEvent);
      expect(onChangeStub.callCount).toBe(1);
      // Expect that the new selection value is false
      expect(onChangeStub.args[0][0]).toBe(false);
    });
  });
});
