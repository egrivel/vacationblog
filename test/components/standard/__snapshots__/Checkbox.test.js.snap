// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/standard/Checkbox #render renders checked 1`] = `
<div>
  <div
    className="formLabel"
  />
  <div
    className="formValue"
  >
    <label
      className="checkbox"
      htmlFor="test-field-id"
      onClick={[Function]}
    >
      <input
        checked={true}
        id="test-field-id"
        onChange={[Function]}
        type="checkbox"
        value="Y"
      />
      Test Label
    </label>
  </div>
</div>
`;

exports[`components/standard/Checkbox #render renders unchecked 1`] = `
<div>
  <div
    className="formLabel"
  />
  <div
    className="formValue"
  >
    <label
      className="checkbox"
      htmlFor="test-field-id"
      onClick={[Function]}
    >
      <input
        id="test-field-id"
        onChange={[Function]}
        type="checkbox"
        value="Y"
      />
      Test Label
    </label>
  </div>
</div>
`;
